Native c10d_functional ops

yifuwang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #110570



This PR introduces a native version of c10d_functional ops. The main goal is to add collective support in AOTInductor and allow collective ops to work in multi-threaded native runtimes.

The native version also incorporated API improvements we wished to implement in Python c10d_functional:

- Removed `ranks` and `group_size` from collective op signatures which were proven to be redundant.
- Use tensor storage as opposed to `void*` to resolve in-flight work.

The native process group registration/resolution mechansim is only used for native c10d_functional in the PR. It will become the single source of truth in upcoming PRs.

The upcoming PRs will implement Inductor/AOTInductor support for c10d_functional, after which native c10d_functional will replace Python c10d_functional.

Commit:
Native c10d_functional ops

This PR introduces a native version of c10d_functional ops. The main goal is to add collective support in AOTInductor and allow collective ops to work in multi-threaded native runtimes.

The native version also incorporated API improvements we wished to implement in Python c10d_functional:

- Removed `ranks` and `group_size` from collective op signatures which were proven to be redundant.
- Use tensor storage as opposed to `void*` to resolve in-flight work.

The native process group registration/resolution mechansim is only used for native c10d_functional in the PR. It will become the single source of truth in upcoming PRs.

The upcoming PRs will implement Inductor/AOTInductor support for c10d_functional, after which native c10d_functional will replace Python c10d_functional.

Differential Revision: [D49932255](https://our.internmc.facebook.com/intern/diff/D49932255/)

[ghstack-poisoned]

Commit:
Update on "Native c10d_functional ops"


This PR introduces a native version of c10d_functional ops. The main goal is to add collective support in AOTInductor and allow collective ops to work in multi-threaded native runtimes.

The native version also incorporated API improvements we wished to implement in Python c10d_functional:

- Removed `ranks` and `group_size` from collective op signatures which were proven to be redundant.
- Use tensor storage as opposed to `void*` to resolve in-flight work.

The native process group registration/resolution mechansim is only used for native c10d_functional in the PR. It will become the single source of truth in upcoming PRs.

The upcoming PRs will implement Inductor/AOTInductor support for c10d_functional, after which native c10d_functional will replace Python c10d_functional.

Differential Revision: [D49932255](https://our.internmc.facebook.com/intern/diff/D49932255/)

[ghstack-poisoned]

wanchaol commented on torch/csrc/distributed/c10d/Functional.cpp:
```diff
@@ -0,0 +1,362 @@
+
+TORCH_LIBRARY(_c10d_functional, m) {
+  m.def(
+      "all_reduce(Tensor input, str reduce_op, str tag) -> Tensor",
```
"tag" is a bit hard for user to understand, wondering if we should do `group_name` instead?

yifuwang replied:
Sounds good!

wanchaol commented on torch/csrc/distributed/c10d/Functional.cpp:
```diff
@@ -0,0 +1,362 @@
+// associate with different process group objects based on the same tag, making
+// the resolution logic context-dependent. While using thread_local might
+// appear as a feasible solution, it falls short during backward passes. This
+// is because an autograd thread cannot be reliably linked to a specific rank.
```
iirc in autograd there's a "driver thread" and worker threads to which really execute the backward nodes. We need this autograd thread id because we don't know which thread we are in during backward, as backward nodes itself might contain collective?

yifuwang replied:
Exactly. In GPU workloads, each autograd worker thread maps to a single device. In theory, we could resolve the rank via the device. However, this doesn't work in CPU workloads where work assignment is arbitrary. The approach in this PR works for both GPU and CPU workloads.

wanchaol commented on test/distributed/test_c10d_functional_native.py:
```diff
@@ -0,0 +1,189 @@
+
+
+@requires_nccl()
+class C10DFunctionalNativeTest(MultiProcessTestCase):
```
do we also need to test these in MultithreadedTestCase to make sure things work in multithreaded context?

wanchaol commented on test/distributed/test_c10d_functional_native.py:
```diff
@@ -0,0 +1,189 @@
+# Owner(s): ["module: c10d"]
+from typing import List
```
nit: shall we merge this test file with the `test_functional_api.py`?

yifuwang replied:
We'll replace the Python c10d_functional with the native version in subsequent PRs. I'll merge the tests when that happens.

Commit:
Update on "Native c10d_functional ops"


This PR introduces a native version of c10d_functional ops. The main goal is to add collective support in AOTInductor and allow collective ops to work in multi-threaded native runtimes.

The native version also incorporated API improvements we wished to implement in Python c10d_functional:

- Removed `ranks` and `group_size` from collective op signatures which were proven to be redundant.
- Use tensor storage as opposed to `void*` to resolve in-flight work.

The native process group registration/resolution mechansim is only used for native c10d_functional in the PR. It will become the single source of truth in upcoming PRs.

The upcoming PRs will implement Inductor/AOTInductor support for c10d_functional, after which native c10d_functional will replace Python c10d_functional.

Differential Revision: [D49932255](https://our.internmc.facebook.com/intern/diff/D49932255/)

[ghstack-poisoned]

Commit:
Update on "Native c10d_functional ops"


This PR introduces a native version of c10d_functional ops. The main goal is to add collective support in AOTInductor and allow collective ops to work in multi-threaded native runtimes.

The native version also incorporated API improvements we wished to implement in Python c10d_functional:

- Removed `ranks` and `group_size` from collective op signatures which were proven to be redundant.
- Use tensor storage as opposed to `void*` to resolve in-flight work.

The native process group registration/resolution mechansim is only used for native c10d_functional in the PR. It will become the single source of truth in upcoming PRs.

The upcoming PRs will implement Inductor/AOTInductor support for c10d_functional, after which native c10d_functional will replace Python c10d_functional.

Differential Revision: [D49932255](https://our.internmc.facebook.com/intern/diff/D49932255/)

[ghstack-poisoned]

Commit:
Update on "Native c10d_functional ops"


This PR introduces a native version of c10d_functional ops. The main goal is to add collective support in AOTInductor and allow collective ops to work in multi-threaded native runtimes.

The native version also incorporated API improvements we wished to implement in Python c10d_functional:

- Removed `ranks` and `group_size` from collective op signatures which were proven to be redundant.
- Use tensor storage as opposed to `void*` to resolve in-flight work.

The native process group registration/resolution mechansim is only used for native c10d_functional in the PR. It will become the single source of truth in upcoming PRs.

The upcoming PRs will implement Inductor/AOTInductor support for c10d_functional, after which native c10d_functional will replace Python c10d_functional.

Differential Revision: [D49932255](https://our.internmc.facebook.com/intern/diff/D49932255/)

[ghstack-poisoned]

Commit:
Update on "Native c10d_functional ops"


This PR introduces a native version of c10d_functional ops. The main goal is to add collective support in AOTInductor and allow collective ops to work in multi-threaded native runtimes.

The native version also incorporated API improvements we wished to implement in Python c10d_functional:

- Removed `ranks` and `group_size` from collective op signatures which were proven to be redundant.
- Use tensor storage as opposed to `void*` to resolve in-flight work.

The native process group registration/resolution mechansim is only used for native c10d_functional in the PR. It will become the single source of truth in upcoming PRs.

The upcoming PRs will implement Inductor/AOTInductor support for c10d_functional, after which native c10d_functional will replace Python c10d_functional.

Differential Revision: [D49932255](https://our.internmc.facebook.com/intern/diff/D49932255/)

[ghstack-poisoned]

wanchaol commented on torch/csrc/distributed/c10d/Functional.cpp:
```diff
@@ -0,0 +1,258 @@
+    {"bxor", c10d::ReduceOp(c10d::ReduceOp::RedOpType::BXOR)},
+    // TODO: support premul_sum
+    // {"premul_sum", c10d::ReduceOp(c10d::ReduceOp::RedOpType::PREMUL_SUM)},
+    {"unused", c10d::ReduceOp(c10d::ReduceOp::RedOpType::UNUSED)}};
```
Oh interesting I didn't know we have this

wanchaol commented on torch/csrc/distributed/c10d/Functional.cpp:
```diff
@@ -0,0 +1,258 @@
+
+std::vector<at::Tensor> all_gather_into_tensor_coalesced(
+    const std::vector<at::Tensor>& inputs,
+    const int64_t group_size,
```
curious why we need `group_size` as a argument, i thought the `group_size` can be resolved after the we call `group  = resolve_process_group` and then use `group.size()`?

yifuwang replied:
I also really don't like the redundancy here.

Since the output shape depends on `group_size`, without it we'd need to query the group info inside meta kernels. This is fine for just-in-time compilation where the group info is available. However, it wouldn't work if we export the program and compile it in another environment.

wanchaol replied:
I see, hmmm if we export the program and compile it in another environment, we still need to somehow get the corresponding `group_name -> pg` setup otherwise when we run it in another environment, the process group resolution would also fail?

yifuwang replied:
If we are only compiling I think we can get away with not having the mapping available (e.g. model transformation + aot_compile for inference). But yeah we need this information eventually at runtime. I'll think about if we want to export this information in a standard way (and use it to create fake pgs for compilation). Will revisit this before making the API public.

wanchaol replied:
Yeah sounds good, let's do it as a follow up. I feel if this is just for the meta function we should be able to hide this inside the meta function/real function op itself by resolving the group and get the group size, as no matter where we run the compiled program we would need to setup some pg.

wanchaol commented on torch/csrc/distributed/c10d/Functional.cpp:
```diff
@@ -0,0 +1,258 @@
+
+at::Tensor all_gather_into_tensor(
+    const at::Tensor& input,
+    const int64_t group_size,
```
ditto: I feel we can get rid of the group_size and infer it from the resolved pg?

wanchaol commented on torch/csrc/distributed/c10d/Functional.cpp:
```diff
@@ -0,0 +1,258 @@
+  return all_gather_into_tensor_coalesced(inputs, group_size, group_name)[0];
+}
+
+at::Tensor allocate_reduce_scatter_output(
```
nit: maybe make these util functions private to distinguish with the collective APIs we defined?

wanchaol commented on torch/distributed/_functional_collectives.py:
```diff
@@ -566,6 +566,27 @@ def _all_to_all_single_meta(input, output_split_sizes, input_split_sizes, tag, r
 
+def _all_gather_into_tensor_native_meta(input, group_size, group_name):
+    shape = list(input.size())
+    shape[0] *= group_size
```
I guess the group size is because of those meta functions? wondering if we can also call resolve_group inside those meta functions

wanchaol commented on torch/distributed/_functional_collectives.py:
```diff
@@ -596,6 +617,15 @@ def _register_ops():
+    _c10_lib_impl.impl("all_gather_into_tensor", _all_gather_into_tensor_native_meta, "Meta")
+    _c10_lib_impl.impl("all_gather_into_tensor_coalesced", _all_gather_into_tensor_coalesced_native_meta, "Meta")
+    _c10_lib_impl.impl("reduce_scatter_tensor", _reduce_scatter_tensor_native_meta, "Meta")
+    _c10_lib_impl.impl("reduce_scatter_tensor_coalesced", _reduce_scatter_tensor_coalesced_native_meta, "Meta")
```
I guess we would merge `_c10d_functional` and `c10d_functional` in follow up PRs?

wanchaol reviewed (approved):
Looks great! Thanks for working on this! I left a few comments inlined, mostly related to whether we need this `group_size` argument to our aten op schema as i imagine we can get rid of that and resolve that group size after the retrival of the pg. Stamp to unblock

yifuwang commented:
@pytorchbot merge

Commit:
Update on "Native c10d_functional ops"




This PR introduces a native version of c10d_functional ops. The main goal is to add collective support in AOTInductor and allow collective ops to work in multi-threaded native runtimes.

The native version also incorporated API improvements we wished to implement in Python c10d_functional:

- Removed `ranks` and `group_size` from collective op signatures which were proven to be redundant.
- Use tensor storage as opposed to `void*` to resolve in-flight work.

The native process group registration/resolution mechansim is only used for native c10d_functional in the PR. It will become the single source of truth in upcoming PRs.

The upcoming PRs will implement Inductor/AOTInductor support for c10d_functional, after which native c10d_functional will replace Python c10d_functional.


Differential Revision: [D49932255](https://our.internmc.facebook.com/intern/diff/D49932255)

[ghstack-poisoned]

yifuwang commented:
@pytorchbot merge
